File: crewai-python/app/pipeline.py
"""CrewAI feature-delivery pipeline."""

from __future__ import annotations

import uuid
from pathlib import Path
from typing import Type

from crewai import Agent, Crew, LLM, Process, Task
from crewai.tools import BaseTool
from pydantic import BaseModel, Field

from app.llm_config import resolve_llm_settings
from app.models import RunResult
from app.phase_log import (
    ensure_prerequisites,
    load_artifacts,
    log_phase_start,
    parse_agents,
    validate_agent_selection,
)
from app.tools import Sandbox

ROOT = Path(__file__).resolve().parent.parent
KNOWLEDGE_DIR = ROOT / "knowledge"
OUTPUT_DIR = ROOT / "output"

SPANISH = (
    "Responde siempre en español. Identificadores de código en inglés cuando sea natural. "
    "Sé concreto y accionable."
)

FRAMEWORK = "crewai"


class _QueryInput(BaseModel):
    query: str = Field(..., description="Search query")


class _PathInput(BaseModel):
    path: str = Field(..., description="Relative path inside the sandbox")


class _WriteInput(BaseModel):
    path: str = Field(..., description="Relative path inside the sandbox")
    content: str = Field(..., description="File contents")


class _PrefixInput(BaseModel):
    prefix: str = Field("", description="Optional subdirectory prefix")


class _ImageInput(BaseModel):
    prompt: str = Field(..., description="Image prompt in English")
    path: str = Field(..., description="Output path e.g. mockup_1.png")


def _make_tools(sandbox: Sandbox) -> list[BaseTool]:
    class SearchKnowledgeTool(BaseTool):
        name: str = "search_knowledge"
        description: str = "Search local engineering knowledge docs"
        args_schema: Type[BaseModel] = _QueryInput

        def _run(self, query: str) -> str:
            return sandbox.search_knowledge(query)

    class WebSearchTool(BaseTool):
        name: str = "web_search"
        description: str = "Search the web via DuckDuckGo (no API key required)"
        args_schema: Type[BaseModel] = _QueryInput

        def _run(self, query: str) -> str:
            return sandbox.web_search(query)

    class WriteFileTool(BaseTool):
        name: str = "write_file"
        description: str = "Write a file inside the run sandbox"
        args_schema: Type[BaseModel] = _WriteInput

        def _run(self, path: str, content: str) -> str:
            return sandbox.write_file(path, content)

    class WriteMermaidTool(BaseTool):
        name: str = "write_mermaid"
        description: str = "Write a Mermaid diagram file under diagrams/"
        args_schema: Type[BaseModel] = _WriteInput

        def _run(self, path: str, content: str) -> str:
            return sandbox.write_mermaid(path, content)

    class GenerateImageTool(BaseTool):
        name: str = "generate_image"
        description: str = "Generate a PNG mockup via OpenAI Images (requires OPENAI_API_KEY)"
        args_schema: Type[BaseModel] = _ImageInput

        def _run(self, prompt: str, path: str) -> str:
            return sandbox.generate_image(prompt, path)

    class ReadFileTool(BaseTool):
        name: str = "read_file"
        description: str = "Read a file from the run sandbox"
        args_schema: Type[BaseModel] = _PathInput

        def _run(self, path: str) -> str:
            return sandbox.read_file(path)

    class ListFilesTool(BaseTool):
        name: str = "list_files"
        description: str = "List files in the run sandbox"
        args_schema: Type[BaseModel] = _PrefixInput

        def _run(self, prefix: str = "") -> str:
            return sandbox.list_files(prefix)

    return [
        SearchKnowledgeTool(),
        WebSearchTool(),
        WriteFileTool(),
        WriteMermaidTool(),
        GenerateImageTool(),
        ReadFileTool(),
        ListFilesTool(),
    ]


def _crew_llm(provider: str | None, model: str | None) -> tuple[LLM, object]:
    settings = resolve_llm_settings(provider, model)
    if settings.provider == "deepseek":
        llm = LLM(
            model=f"openai/{settings.model}",
            api_key=settings.api_key,
            base_url=settings.base_url,
            temperature=0.2,
        )
    else:
        llm = LLM(
            model=f"openai/{settings.model}",
            api_key=settings.api_key,
            temperature=0.2,
        )
    return llm, settings


def _make_agents(llm: LLM, tools: list[BaseTool]) -> dict[str, Agent]:
    return {
        "researcher": Agent(
            role="Researcher",
            goal="Investigar en web y knowledge local; producir research.md",
            backstory=f"Analista que sintetiza fuentes externas e internas. {SPANISH}",
            llm=llm,
            tools=tools,
            verbose=True,
            allow_delegation=False,
        ),
        "planner": Agent(
            role="Planner",
            goal="Descomponer la feature en criterios de aceptación y tareas ordenadas",
            backstory=f"Tech lead enfocado en alcance y riesgos. {SPANISH}",
            llm=llm,
            tools=tools,
            verbose=True,
            allow_delegation=False,
        ),
        "designer": Agent(
            role="Designer",
            goal="Proponer arquitectura, APIs y modelo de datos",
            backstory=f"Arquitecto de software pragmático. {SPANISH}",
            llm=llm,
            tools=tools,
            verbose=True,
            allow_delegation=False,
        ),
        "diagrammer": Agent(
            role="Diagrammer",
            goal="Crear diagramas Mermaid (flowchart) de arquitectura y flujo temporal",
            backstory=(
                f"Especialista en visualización técnica. Solo flowchart/graph "
                f"(nunca sequenceDiagram). {SPANISH}"
            ),
            llm=llm,
            tools=tools,
            verbose=True,
            allow_delegation=False,
        ),
        "illustrator": Agent(
            role="Illustrator",
            goal="Generar imágenes o mockups PNG según el pedido",
            backstory=(
                f"Ilustrador que produce assets PNG (imágenes generales o mockups UI). {SPANISH}"
            ),
            llm=llm,
            tools=tools,
            verbose=True,
            allow_delegation=False,
        ),
        "coder": Agent(
            role="Coder",
            goal="Implementar la feature en el sandbox usando write_file bajo src/",
            backstory=f"Senior engineer que entrega código claro y mínimo viable. {SPANISH}",
            llm=llm,
            tools=tools,
            verbose=True,
            allow_delegation=False,
        ),
        "reviewer": Agent(
            role="Reviewer",
            goal="Revisar plan, diseño y código; emitir veredicto",
            backstory=f"Staff engineer estricto en seguridad y aceptación. {SPANISH}",
            llm=llm,
            tools=tools,
            verbose=True,
            allow_delegation=False,
        ),
    }


def _build_task(
    phase: str,
    request: str,
    agent_map: dict[str, Agent],
    plan: str,
    design: str,
    research: str,
) -> Task:
    if phase == "researcher":
        return Task(
            description=(
                f"Feature request:\n{request}\n\n"
                "Usa web_search y search_knowledge. Escribe research.md con write_file: "
                "fuentes citadas, resumen e implicaciones."
            ),
            expected_output="Contenido de research.md en español",
            agent=agent_map["researcher"],
        )
    if phase == "planner":
        return Task(
            description=(
                f"Feature request:\n{request}\n\nResearch:\n{research}\n\n"
                "Usa search_knowledge si ayuda. Escribe el plan completo en plan.md con write_file. "
                "Incluye criterios de aceptación, tareas, riesgos y fuera de alcance."
            ),
            expected_output="Contenido de plan.md en español",
            agent=agent_map["planner"],
        )
    if phase == "designer":
        return Task(
            description=(
                f"Feature request:\n{request}\n\nPlan existente:\n{plan}\n\n"
                "Produce design.md (write_file) con componentes, endpoints, "
                "modelo de datos y trade-offs. Usa search_knowledge sobre api design si hace falta."
            ),
            expected_output="Contenido de design.md en español",
            agent=agent_map["designer"],
        )
    if phase == "diagrammer":
        return Task(
            description=(
                f"Feature request:\n{request}\n\nPlan:\n{plan}\n\nDesign:\n{design}\n\n"
                "Usa write_mermaid para diagrams/architecture.mmd y diagrams/sequence.mmd. "
                "REGLA: cada archivo debe empezar con 'flowchart TD' o 'flowchart LR' (o 'graph TD'). "
                "NO uses sequenceDiagram ni classDiagram. "
                "sequence.mmd = flujo temporal con flowchart; architecture.mmd = componentes."
            ),
            expected_output="Diagramas Mermaid flowchart bajo diagrams/",
            agent=agent_map["diagrammer"],
        )
    if phase == "illustrator":
        return Task(
            description=(
                f"Request:\n{request}\n\nResearch:\n{research}\n\nDesign:\n{design}\n\n"
                "Usa generate_image para 1-2 imágenes en assets/ (prompts en inglés). "
                "Si hay Design de producto, prioriza mockups UI; si no, genera la imagen "
                "pedida usando Research/Request."
            ),
            expected_output="PNG bajo assets/",
            agent=agent_map["illustrator"],
        )
    if phase == "coder":
        return Task(
            description=(
                f"Feature request:\n{request}\n\nPlan:\n{plan}\n\nDesign:\n{design}\n\n"
                "Implementa la feature. Crea archivos bajo src/ con write_file "
                "(p.ej. src/main.py, src/auth.py, src/README.md). Código funcional mínimo."
            ),
            expected_output="Lista de archivos creados bajo src/",
            agent=agent_map["coder"],
        )
    return Task(
        description=(
            f"Feature request:\n{request}\n\n"
            "Lee plan.md, design.md y el código con read_file/list_files. "
            "Consulta el checklist de review en knowledge. Escribe review.md con hallazgos "
            "y veredicto approve|request_changes|comment."
        ),
        expected_output="Contenido de review.md en español",
        agent=agent_map["reviewer"],
    )


def run_feature_delivery(
    request: str,
    *,
    provider: str | None = None,
    model: str | None = None,
    run_id: str | None = None,
    agents: str | list[str] | None = None,
) -> RunResult:
    selected = parse_agents(agents)
    run_id = run_id or uuid.uuid4().hex[:10]
    output_dir = OUTPUT_DIR / run_id
    sandbox = Sandbox(output_dir, KNOWLEDGE_DIR)
    validate_agent_selection(selected, sandbox)
    tools = _make_tools(sandbox)
    llm, settings = _crew_llm(provider, model)
    agent_map = _make_agents(llm, tools)

    artifacts = load_artifacts(sandbox)
    plan = str(artifacts["plan"])
    design = str(artifacts["design"])
    research = str(artifacts["research"])

    total = len(selected)
    for index, phase in enumerate(selected, start=1):
        ensure_prerequisites(phase, sandbox)
        log_phase_start(phase, index, total, run_id, FRAMEWORK)

        task = _build_task(phase, request, agent_map, plan, design, research)
        crew = Crew(
            agents=[agent_map[phase]],
            tasks=[task],
            process=Process.sequential,
            verbose=True,
        )
        crew.kickoff()

        artifacts = load_artifacts(sandbox)
        plan = str(artifacts["plan"])
        design = str(artifacts["design"])
        research = str(artifacts["research"])

    plan_text = sandbox.read_file("plan.md")
    design_text = sandbox.read_file("design.md")
    review_text = sandbox.read_file("review.md")
    research_text = sandbox.read_file("research.md")
    if plan_text.startswith("ERROR"):
        plan_text = plan
    if design_text.startswith("ERROR"):
        design_text = design
    if review_text.startswith("ERROR"):
        review_text = ""
    if research_text.startswith("ERROR"):
        research_text = research

    result = RunResult(
        run_id=run_id,
        output_dir=str(output_dir),
        request=request,
        research=research_text,
        plan=plan_text,
        design=design_text,
        review=review_text,
        files=sandbox.list_written_files(),
        diagrams=list(artifacts["diagrams"]),  # type: ignore[arg-type]
        assets=list(artifacts["assets"]),  # type: ignore[arg-type]
        provider=settings.provider,
        model=settings.model,
    )
    result.write_summary()
    result.announce()
    return result

File: mastra-typescript/src/pipeline.ts
import "dotenv/config";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { randomBytes } from "node:crypto";
import { createOpenAI } from "@ai-sdk/openai";
import { Agent } from "@mastra/core/agent";
import { createTool } from "@mastra/core/tools";
import { z } from "zod";
import { resolveLlmSettings } from "./llmConfig.js";
import { Sandbox } from "./tools.js";
import type { RunResult } from "./models.js";
import { announceResults, writeSummary } from "./models.js";
import {
  ensurePrerequisites,
  loadArtifacts,
  logPhaseStart,
  parseAgents,
  validateAgentSelection,
  type Phase,
} from "./phaseLog.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
export const ROOT = path.resolve(__dirname, "..");
export const KNOWLEDGE_DIR = path.join(ROOT, "knowledge");
export const OUTPUT_DIR = path.join(ROOT, "output");

const SPANISH =
  "Responde siempre en español. Identificadores de código en inglés cuando sea natural.";

const FRAMEWORK = "mastra";

export type RunOptions = {
  provider?: string;
  model?: string;
  runId?: string;
  agents?: string | string[];
};

function buildModel(provider?: string, model?: string) {
  const settings = resolveLlmSettings(provider, model);
  const openai = createOpenAI({
    apiKey: settings.apiKey,
    baseURL: settings.baseUrl,
  });
  return { model: openai(settings.model), settings };
}

function buildTools(sandbox: Sandbox) {
  const searchKnowledge = createTool({
    id: "search_knowledge",
    description: "Search local engineering knowledge docs",
    inputSchema: z.object({ query: z.string() }),
    execute: async ({ context }) => sandbox.searchKnowledge(context.query),
  });
  const webSearch = createTool({
    id: "web_search",
    description: "Search the web via DuckDuckGo (no API key required)",
    inputSchema: z.object({ query: z.string() }),
    execute: async ({ context }) => sandbox.webSearch(context.query),
  });
  const writeFile = createTool({
    id: "write_file",
    description: "Write a file inside the run sandbox",
    inputSchema: z.object({ path: z.string(), content: z.string() }),
    execute: async ({ context }) =>
      sandbox.writeFile(context.path, context.content),
  });
  const writeMermaid = createTool({
    id: "write_mermaid",
    description: "Write a Mermaid diagram file under diagrams/",
    inputSchema: z.object({ path: z.string(), content: z.string() }),
    execute: async ({ context }) =>
      sandbox.writeMermaid(context.path, context.content),
  });
  const generateImage = createTool({
    id: "generate_image",
    description: "Generate a PNG mockup via OpenAI Images (requires OPENAI_API_KEY)",
    inputSchema: z.object({ prompt: z.string(), path: z.string() }),
    execute: async ({ context }) =>
      sandbox.generateImage(context.prompt, context.path),
  });
  const readFile = createTool({
    id: "read_file",
    description: "Read a file from the run sandbox",
    inputSchema: z.object({ path: z.string() }),
    execute: async ({ context }) => sandbox.readFile(context.path),
  });
  const listFiles = createTool({
    id: "list_files",
    description: "List files in the run sandbox",
    inputSchema: z.object({ prefix: z.string().optional() }),
    execute: async ({ context }) => sandbox.listFiles(context.prefix ?? ""),
  });
  return {
    searchKnowledge,
    webSearch,
    writeFile,
    writeMermaid,
    generateImage,
    readFile,
    listFiles,
  };
}

type PhaseContext = {
  request: string;
  research: string;
  plan: string;
  design: string;
};

function buildPhaseRunners(
  sandbox: Sandbox,
  model: ReturnType<ReturnType<typeof createOpenAI>>,
  tools: ReturnType<typeof buildTools>,
) {
  const researcher = new Agent({
    name: "Researcher",
    instructions: `${SPANISH} Eres el Researcher. Usa web_search y search_knowledge. Escribe research.md con write_file.`,
    model,
    tools,
  });
  const planner = new Agent({
    name: "Planner",
    instructions: `${SPANISH} Eres el Planner. Usa search_knowledge si ayuda. Escribe plan.md con write_file.`,
    model,
    tools,
  });
  const designer = new Agent({
    name: "Designer",
    instructions: `${SPANISH} Eres el Designer. Escribe design.md con write_file (componentes, APIs, datos).`,
    model,
    tools,
  });
  const diagrammer = new Agent({
    name: "Diagrammer",
    instructions: `${SPANISH} Eres el Diagrammer. Usa write_mermaid para diagrams/architecture.mmd y diagrams/sequence.mmd. REGLA: cada archivo debe empezar con 'flowchart TD' o 'flowchart LR' (o 'graph TD'). NO uses sequenceDiagram ni classDiagram. sequence.mmd = flujo temporal con flowchart; architecture.mmd = componentes.`,
    model,
    tools,
  });
  const illustrator = new Agent({
    name: "Illustrator",
    instructions: `${SPANISH} Eres el Illustrator. Usa generate_image para 1-2 imágenes en assets/. Si hay Design de producto, prioriza mockups UI; si no, genera la imagen pedida.`,
    model,
    tools,
  });
  const coder = new Agent({
    name: "Coder",
    instructions: `${SPANISH} Eres el Coder. Implementa bajo src/ con write_file.`,
    model,
    tools,
  });
  const reviewer = new Agent({
    name: "Reviewer",
    instructions: `${SPANISH} Eres el Reviewer. Lee artefactos con tools y escribe review.md con veredicto.`,
    model,
    tools,
  });

  const runners: Record<
    Phase,
    (ctx: PhaseContext) => Promise<{ research: string; plan: string; design: string; review: string }>
  > = {
    researcher: async (ctx) => {
      const res = await researcher.generate(
        `Feature request:\n${ctx.request}\nEscribe research.md completo.`,
      );
      const researchFile = sandbox.readFile("research.md");
      const research = researchFile.startsWith("ERROR")
        ? String(res.text ?? "")
        : researchFile;
      return { research, plan: ctx.plan, design: ctx.design, review: "" };
    },
    planner: async (ctx) => {
      const res = await planner.generate(
        `Feature request:\n${ctx.request}\n\nResearch:\n${ctx.research}\nEscribe plan.md completo.`,
      );
      const planFile = sandbox.readFile("plan.md");
      const plan = planFile.startsWith("ERROR") ? String(res.text ?? "") : planFile;
      return { research: ctx.research, plan, design: ctx.design, review: "" };
    },
    designer: async (ctx) => {
      const res = await designer.generate(
        `Feature:\n${ctx.request}\n\nPlan:\n${ctx.plan}\nEscribe design.md.`,
      );
      const designFile = sandbox.readFile("design.md");
      const design = designFile.startsWith("ERROR")
        ? String(res.text ?? "")
        : designFile;
      return { research: ctx.research, plan: ctx.plan, design, review: "" };
    },
    diagrammer: async (ctx) => {
      await diagrammer.generate(
        `Feature:\n${ctx.request}\n\nPlan:\n${ctx.plan}\n\nDesign:\n${ctx.design}\n` +
          `Crea diagrams/architecture.mmd y diagrams/sequence.mmd con write_mermaid. ` +
          `Usa solo flowchart TD/LR o graph TD (nunca sequenceDiagram).`,
      );
      return { research: ctx.research, plan: ctx.plan, design: ctx.design, review: "" };
    },
    illustrator: async (ctx) => {
      await illustrator.generate(
        `Request:\n${ctx.request}\n\nResearch:\n${ctx.research}\n\nDesign:\n${ctx.design}\n` +
          `Usa generate_image para 1-2 imágenes en assets/ (prompts en inglés). ` +
          `Si hay Design de producto, prioriza mockups UI; si no, genera la imagen ` +
          `pedida usando Research/Request.`,
      );
      return { research: ctx.research, plan: ctx.plan, design: ctx.design, review: "" };
    },
    coder: async (ctx) => {
      await coder.generate(
        `Feature:\n${ctx.request}\n\nPlan:\n${ctx.plan}\n\nDesign:\n${ctx.design}\n` +
          `Implementa archivos bajo src/ con write_file.`,
      );
      return { research: ctx.research, plan: ctx.plan, design: ctx.design, review: "" };
    },
    reviewer: async (ctx) => {
      const res = await reviewer.generate(
        `Feature:\n${ctx.request}\nLee plan.md, design.md y src/. Escribe review.md ` +
          `con hallazgos y veredicto approve|request_changes|comment. Usa search_knowledge para checklist.`,
      );
      const reviewFile = sandbox.readFile("review.md");
      const review = reviewFile.startsWith("ERROR")
        ? String(res.text ?? "")
        : reviewFile;
      return { research: ctx.research, plan: ctx.plan, design: ctx.design, review };
    },
  };

  return runners;
}

export async function runFeatureDelivery(
  request: string,
  options: RunOptions = {},
): Promise<RunResult> {
  const selected = parseAgents(options.agents);
  const { model, settings } = buildModel(options.provider, options.model);
  const runId = options.runId ?? randomBytes(5).toString("hex");
  const outputDir = path.join(OUTPUT_DIR, runId);
  const sandbox = new Sandbox(outputDir, KNOWLEDGE_DIR);
  validateAgentSelection(selected, sandbox);
  const tools = buildTools(sandbox);
  const runners = buildPhaseRunners(sandbox, model, tools);

  const artifacts = loadArtifacts(sandbox);
  let ctx: PhaseContext = {
    request,
    research: artifacts.research,
    plan: artifacts.plan,
    design: artifacts.design,
  };
  let review = artifacts.review;

  const total = selected.length;
  for (let index = 0; index < selected.length; index++) {
    const phase = selected[index];
    ensurePrerequisites(phase, sandbox);
    logPhaseStart(phase, index + 1, total, runId, FRAMEWORK);
    const result = await runners[phase](ctx);
    ctx = {
      request,
      research: result.research,
      plan: result.plan,
      design: result.design,
    };
    if (result.review) review = result.review;
  }

  const clean = (v: string) => (v?.startsWith("ERROR") ? "" : v || "");
  const finalArtifacts = loadArtifacts(sandbox);

  const runResult: RunResult = {
    run_id: runId,
    output_dir: outputDir,
    request,
    research: clean(ctx.research || sandbox.readFile("research.md")),
    plan: clean(ctx.plan || sandbox.readFile("plan.md")),
    design: clean(ctx.design || sandbox.readFile("design.md")),
    review: clean(review || sandbox.readFile("review.md")),
    files: sandbox.listWrittenFiles(),
    diagrams: finalArtifacts.diagrams,
    assets: finalArtifacts.assets,
    provider: settings.provider,
    model: settings.model,
  };
  writeSummary(runResult);
  announceResults(runResult);
  return runResult;
}
